text/0000-custom-element.md

```diff
@@ -0,0 +1,59 @@
+- Start Date: 2019-07-05
+- RFC PR: (leave this empty)
+- Lightning Web Component Issue: (leave this empty)
+
+# Summary
+
+Facilitate the registration of LWC Constructor as Custom Elements.
+
+# Motivation
+
+As today, the way to generate a qualifying Custom Element from a LWC Constructor, is via an experimental API: `buildCustomElementConstructor()`. Since this is a primary use case for LWC, we want to make it more ergonomic, and simpler. Additionally, the experimental API allow you to generate more than one Custom Element per LWC Constructors, which does not align with the web components semantics.
+
+# Goals
+
+1. Provide a dead simple way to register a custom element from a LWC Constructor.
+2. Preserve the web components semantics.
+
+# Proposal
+
+This proposal provides a high-level API (an abstraction layer) for authors to access the custom element associate to a LWC Constructor:
+
+```js
+import XFoo from 'x/foo';
+customElements.define('x-foo', XFoo.CustomElement);
+// using tagName `x-foo`:
+document.body.appendChild(document.createElement('x-foo'));
+```
+
+Pros:
+* it is extremely simple.
+* it reads very nicely in plain english.
+* it preserves the semantics of web components in a way that you can't bend them.
+
+Cons:
+* it can't be statically analyzed (but since we don't use this in platform, that should be ok).
+* it does not provide a way for you to control whether or not you want open or closed shadowRoot (this is fine because consumers should not be in control)
+
+## Semantics and Invariants
+
+* There is a 1-1 mapping between LWC Constructors and the corresponding Custom Element.
+* Claiming the Custom Element for the abstract `LightningElement` must be disallow.
+
+## Implementation Details
+
+`LightningElement.CustomElement` is a static getter that can be accessed by any Constructor extending `LightningElement`. This getter can rely on the `this` value, which points to the sub class. This can be used in a Map to link the LWC to the Custom Element. This is very similar to the trick we use in  `LightningContextElement.Provider` for context.
+
+As a consequence of this change, the existing experimental API can be removed.
+
+# Adoption strategy
+
+This is a brand new feature, we just need to document it.
+
+# How we teach this
+
+* The consumer aspect of it is very straight forward considering that they don't have to learn a new API, just a static field name in LWC Constructors.
+
+# Unresolved questions
+
+* How can users control the shadowRoot's mode?
```

Comment:
As you stated above, the consumer should not decide what is the shadowRoot mode. I don't think this statement should be included in the _Unresolved questions_ section.

Reply:
Agree with you, we should resolve it!

In today web components, you can decide the shadowRoot mode to use per component. In the current implementation we have it per component, passing it to the `buildCustomElementConstructor`.

As of [today implementation](https://github.com/salesforce/lwc/pull/1395) of this RFC, all LWC custom elements shadowRoot mode is [open](https://github.com/salesforce/lwc/pull/1395/files#diff-863e83caa80f932b8cd147e027dcc89bR18). It can't be set to close.

Lets decide:
- Should we enforce all LWC custom elements to be in open or close mode? Should we provide a mechanism on how to set it?
- Do we want an LWC custom element to specify the shadowRoot mode? Ex: via static property on the constructor.

cc/ @caridy

Reply:
Ia agreed with @pmdartus this question can be removed from here. It is a decision of the developer or the platform, not the consumer of the component.